### ISSUE TITLE
Pub/Sub component with Session enabled Azure Service Bus Topics do not deliver messages in sequence

### ISSUE THREAD

**Issue by jpiquot** [dapr/components-contrib#3281]
## Expected Behavior

When sessions are enabled, a Pub/Sub component subscription should receive messages in sequential order.

![image](https://github.com/dapr/components-contrib/assets/6775094/a5009b6a-6a32-4987-80a7-71b6e336f8d5)

## Actual Behavior

Messages are being received out of sequence, with the next message arriving before the previous one has been received.

![image](https://github.com/dapr/components-contrib/assets/6775094/98864be6-5fcf-41b2-89fd-974bf13ec85d)

## Steps to Reproduce the Problem

- Create a DAPR application.
- Configure a pub/sub component to connect to an Azure Service Bus Topic.

```
[ApiController]
public class PartiesCommandsController(
    ConventionNamingCommandProcessor commandProcessor,
    IHostEnvironment hostEnvironment,
    ILogger<PartiesCommandsController> logger) : CommandSubmissionController(commandProcessor, hostEnvironment, logger)
{
    /// <summary>
    /// Handle external systems commands as an asynchronous operation.
    /// </summary>
    /// <param name="commandState">State of the command.</param>
    /// <returns>A Task&lt;ActionResult&gt; representing the asynchronous operation.</returns>
    [CustomerCommandsBusTopic]
    [TopicMetadata("requireSessions", "true")]
    [TopicMetadata("sessionIdleTimeoutInSec ", "2")]
    [TopicMetadata("maxConcurrentSessions", "8")]
    [HttpPost("/handle-customer-commands")]
    public async Task<ActionResult> SubmitCustomerCommandsAsync(CommandState commandState)
        => await HandleCommandAsync(
                commandState,
                Customer.GetAggregateName(),
                CancellationToken.None)
            .ConfigureAwait(false);
}
```
I observe the below setup, but it appears to be globally applied to the topic. This configuration might not support concurrent session handling if it is set to 1.

> - maxActiveMessages : Defines the maximum number of messages to be processing or in the buffer at once. This should be at least as big as the maximum concurrent handlers. 
> - maxConcurrentHandlers : Defines the maximum number of concurrent message handlers.

![image](https://github.com/dapr/components-contrib/assets/6775094/46a61301-b4bc-4c38-8d6b-aee2d4a2aaed)

## Release Note

RELEASE NOTE: Fix session enabled topics should send messages in sequence

**Comment by berndverst** [comment]
By design messages are received for the given session first, unless within `sessionIdleTimeoutInSec` window there are no messages available. At which point the next session will be tried.

Up to `maxConcurrentSessions` sessions will be tried in parallel within a given sidecar.

If you have multiple sidecars this of course further adds to the parallelization.

Just like in the image you provided, session 1, 2 and 3 would be processessing / receiving messages in parallel - assuming you have at least 3 `maxConcurrentSessions`. There is no guarantee made by Dapr for which session will start first. And there is no such thing as perfect timing between these sessions - each session is a different goroutine and the timing of those can vary. But for each goroutine, as long as the session as new messages `sessionIdleTimeoutInSec` apart, then the messages within that session are received in order.

Absolute ordering across multiple sessions is impossible and not a support use case.

**Comment by berndverst** [comment]
You may need to reduce the number of sidecars to avoid processing the same session multiple times in parallel.

**Comment by jpiquot** [comment]
@berndverst Thank you for your responses. My primary concern is not the sequence in which the DAPR sidecar processes the sessions. When performing incremental updates for customers, it's crucial to ensure that all messages related to a specific customer (using the customer ID as the session ID) are processed sequentially. If a systemic error occurs in my application, there might be several messages with a retry count greater than zero. However, this would occur only once per session. As shown in the screenshot, multiple messages from the same session have a retry count exceeding zero. The application operates in an Azure Container App with a single sidecar (Scale 1-1).

![image](https://github.com/dapr/components-contrib/assets/6775094/b760f894-1c98-45be-b97d-7275f0d72b08)

**Comment by jjcollinge** [comment]
@jpiquot thanks for raising the issue, happy to try and help.

Based on what you've shown can you just clarify that your application is always erroring and never returning a successful response?

As I can see the messages in your subscription in the message browser I am assuming there is no dead letter setup - is this correct?

Each message appears to be delivered 9 times. This does not mean it has been sent to your app 9 times. Rather it means we have taken the message lock and consumed it 9 times. Dapr resiliency may increase the time we hold the lock whilst retrying the call to the app which could expire and increase the delivery count. However, this should not allow a consumer to take a lock on next message in the session sequence. Do you have any idea why we are consistently seeing 9 delivery count?

**Comment by jpiquot** [comment]
@jjcollinge thanks for your answer.

Messages at the top of the queue are retried up to ten times, in line with the queue's settings for a maximum of 10 retries, before being moved to the dead letter queue, which is currently empty since I have manually cleared it.

**Comment by jjcollinge** [comment]
Thanks for the context.

So just to clarify my understanding. The subscriber is intentionally returning an error on each message that is delivered. This causes the message delivery to fail up to the maximum of 10 times. Service Bus then removes that message from the subscription and moves it to the dead letter queue. The subscriber then takes the next message in the session and continues the cycle. From what I can see in your message explorer the sequence numbers looks to be in order. The reason for many messages having a delivery attempt is due to the message being dropped after 10 attempts and the consumer moving to the next one. Which bit of the behaviour are you not expecting?

**Comment by jpiquot** [comment]
In the list, multiple messages within the same session have an incremented retry count (none are dropped yet). This contradicts the idea of a FIFO message delivery. If message one has retry counts, message two should not have any retries.

The session pattern's primary goal in pub/sub systems is to enable parallel processing of message collections while ensuring that messages within a single collection are delivered in a serialized, FIFO sequence. This ensures that each message is processed in the order it was sent, maintaining the integrity of the message sequence (important for DDD aggregate events).

Moreover, as evident in the Azure Service Bus message list, more than one message has been delivered within the same session, each with a retry count. If the message delivery were functioning correctly as per the FIFO principle, we wouldn't observe multiple messages with incremented retry counts in the same session.

This situation occurs even with a single active sidecar. If there were multiple pods, I would expect the sessions to be distributed across them, but still, each message in a session should be delivered in sequence to the same pod.

**Comment by jjcollinge** [comment]
Ok sorry I am away from my laptop for the holidays so only viewing your image on my phone. I mistakenly thought the messages were being dead lettered in order but now I understand what you mean. I'll have to try and reproduce this when I'm back in the new year.

**Comment by jpiquot** [comment]
Thanks for your reply. Yes sessions are required in the subscription.
![image](https://github.com/dapr/components-contrib/assets/6775094/0bee3361-60d9-4874-9698-3e196b527897)

**Comment by jjcollinge** [comment]
@jpiquot I'm struggling to repro this in testing... I've added a new certification test that publishes 100 messages to 3 sessions concurrently and ensures that with max concurrent sessions set to 8 on 2 subscribing replicas we receive each session in order which is working as expected. Looking back at what you've explained... your app deliberately returns an error when it receives a message. You said you see multiple messages within the session with a delivery count greater than 0 but less than 9 at the same time, correct? Are you able to validate that your app is actually receiving requests out of order just to rule out that this is some eventual consistency in the Service Bus explorer. I'll continue trying to reproduce this with a test in the meantime.

**Comment by jpiquot** [comment]
@jjcollinge thanks for your help. The issue may indeed stem from a specific failure condition in our application. Initially, I suspected the root cause to be related to message sequencing. Observing multiple messages in the same session with a retry count seemed to confirm this.

We identified a problem in the our command handler, which was issuing some events with an invalid session ID (excessively long), thereby creating an error condition.

To monitor this, I need to implement additional state management in the command processor. This will involve storing the message sequence and ensuring that we never receive a message with a sequence number lower than its predecessor. This change will aid in identifying any issues related to message order.
